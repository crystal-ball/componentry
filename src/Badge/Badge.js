import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

export default function Badge(props) {
  const { color, rounded, ...rest } = { ...useTheme('Badge'), ...props }

  return elem({
    elemClassName: {
      badge: true,
      [`badge-${color}`]: color,
      'badge-rounded': rounded,
    },
    ...rest,
  })
}
Badge.displayName = 'Badge'
