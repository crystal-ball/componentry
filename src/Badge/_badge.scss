/* ~componentry/src/Badge/badge */

/**
 * --- Base class
 * Requires one of the contextual, color modifier classes for `color` and
 * `background-color`.
 */
.ðŸ…²-badge {
  display: inline-block;
  padding: $badge-padding-y $badge-padding-x;
  font-size: $badge-font-size;
  font-weight: $badge-font-weight;
  line-height: 1;
  text-align: center;
  white-space: nowrap;
  vertical-align: baseline;
  @include border-radius($badge-border-radius);

  // Empty badges collapse automatically
  &:empty {
    display: none;
  }
}

// Quick fix for badges in buttons
.ðŸ…²-button .ðŸ…²-badge {
  position: relative;
  top: -1px;
}

/**
 * --- Rounded badges
 * Creates a badge with rounded corners and extra padding
 */
.badge-rounded {
  padding-right: $badge-rounded-padding-x;
  padding-left: $badge-rounded-padding-x;
  border-radius: 1000px !important;
}

/**
 * --- Colors
 * Contextual variations (linked badges get darker on :hover).
 */
@mixin badge-variant($bg) {
  color: color-yiq($bg);
  background-color: $bg;

  &[href] {
    @include hover-focus {
      color: color-yiq($bg);
      text-decoration: none;
      background-color: darken($bg, 10%);
    }
  }
}

@each $color, $value in $theme-colors {
  .badge-color-#{$color} {
    @include badge-variant($value);
  }
}
