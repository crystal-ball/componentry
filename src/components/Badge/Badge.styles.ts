import { type CSSProperties } from 'react'
import { Theme } from '../../theme/theme'

//                                         <Badge /> styles
// --------------------------------------------------------

export const badgeStyles = (theme: Theme): BadgeStyles => ({
  // BASE
  '.C9Y-Badge-base': {
    display: 'inline-flex',
    alignItems: 'center',
    whiteSpace: 'nowrap',
  },

  // VARIANTS
  '.C9Y-Badge-filled': {
    padding: '4px 8px',
    color: theme.colors.inverse,
    backgroundColor: theme.colors.gray[700],
    fontWeight: theme.fontWeight.bold,
    borderRadius: theme.borderRadius.DEFAULT,
    fontSize: theme.fontSize.sm,
    lineHeight: 1,
    // ðŸ’¡ Use em with font-size and padding to auto-scale with text
  },

  // SIZES
  '.C9Y-Badge-smallSize': {
    fontSize: theme.fontSize.sm,
    padding: '2px 6px',
  },
  '.C9Y-Badge-largeSize': {
    fontSize: theme.fontSize.body,
    padding: '6px 12px',
  },
})

export interface BadgeStyles {
  '.C9Y-Badge-base': CSSProperties
  '.C9Y-Badge-filled': CSSProperties
  '.C9Y-Badge-smallSize': CSSProperties
  '.C9Y-Badge-largeSize': CSSProperties
}
