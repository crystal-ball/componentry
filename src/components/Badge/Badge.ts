import { useTheme } from '../Theme/Theme'
import { element } from '../../utils/element-creator'
import { type ComponentBaseProps } from '../../utils/base-types'

interface BadgeProps extends ComponentBaseProps<'div'> {
  /** Variant color */
  color?: 'primary'
  /** Display variant */
  variant?: 'filled'
}

/**
 * [Badge component üìù](https://componentry.design/components/badge)
 * @experimental
 */
export const Badge: React.FC<BadgeProps> = (props) => {
  const {
    color,
    variant = 'filled',
    ...rest
  } = {
    ...useTheme<BadgeProps>('Badge'),
    ...props,
  }

  return element({
    componentCx: [
      `C9Y-Badge-base C9Y-Badge-${variant}`,
      { [`C9Y-Badge-${color}Color`]: color },
    ],
    ...rest,
  })
}
Badge.displayName = 'Badge'
