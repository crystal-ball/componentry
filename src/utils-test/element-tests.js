/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
