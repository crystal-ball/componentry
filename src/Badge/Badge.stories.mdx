import { Badge } from './Badge'
import { colors } from '../storybook-resources'

<Meta title='Components/Badge' component={Badge} />

<DocsTitle title='Badges' experimental />

Badges provide status information

<Badge color='primary'>Badge</Badge>

---

## Colors

<div className='d-flex justify-content-center'>
  {colors.map((color) => (
    <div className='px-1' key={color}>
      <Badge color={color}>Badge</Badge>
    </div>
  ))}
</div>

## Rounded badges

Pass the `rounded` prop to make a badge rounded:

<Badge color='primary' rounded>
  Badge
</Badge>

## Routing with badges

<Badge as='a' color='primary' href='http://componentry.design'>
  Componentry
</Badge>
