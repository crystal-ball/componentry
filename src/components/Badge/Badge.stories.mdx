import { Meta, Story } from '@storybook/addon-docs'

import { Text } from '../Text/Text'
import { Badge } from './Badge'

<Meta title='Components/Badge' component={Badge} />

# Badge

Badges provide status information.

<Story name='Badge'>
  <Badge>Badge</Badge>
</Story>

## Sizes

Pass a `size` to set a badge size:

<div className='flex flex-row items-center gap-4'>
  <Badge size='small'>Badge</Badge>
  <Badge>Badge</Badge>
  <Badge size='large'>Badge</Badge>
</div>
