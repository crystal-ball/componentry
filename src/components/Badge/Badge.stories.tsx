import type { Meta, StoryObj } from '@storybook/react'

import { Badge } from './Badge'

const meta: Meta<typeof Badge> = {
  component: Badge,
}

export default meta
type Story = StoryObj<typeof Badge>

export const Primary: Story = {
  args: {
    variant: 'filled',
    color: 'primary',
    children: 'Badge',
  },
}

export const Sizes: Story = {
  render: () => (
    <>
      <Badge size='small'>Badge</Badge>
      <Badge>Badge</Badge>
      <Badge size='large'>Badge</Badge>
    </>
  ),
}
