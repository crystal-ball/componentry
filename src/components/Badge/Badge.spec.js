import React from 'react'
import { render, screen } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { Badge } from './Badge'

describe('<Badge />', () => {
  elementTests(Badge)

  it('When color is passed, then badge-color className is rendered', () => {
    render(<Badge color='primary'>Badge</Badge>)

    expect(screen.getByText('Badge')).toHaveClass('badge-primary')
  })

  it('When variant is rounded, variant className is rendered', () => {
    render(<Badge variant='rounded'>Badge</Badge>)

    expect(screen.getByText('Badge')).toHaveClass('badge-rounded')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Badge /> snapshots', () => {
  it('renders correctly', () => {
    render(
      <Badge variant='rounded' color='primary' data-testid='badge'>
        Badge
      </Badge>,
    )

    expect(screen.getByTestId('badge')).toMatchSnapshot()
  })
})
