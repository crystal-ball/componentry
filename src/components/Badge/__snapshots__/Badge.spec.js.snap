// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> snapshots renders correctly 1`] = `
<div
  class="C9Y-Badge-base C9Y-Badge-filled C9Y-Badge-primaryColor"
  data-testid="badge"
>
  Badge
</div>
`;
